packages/docusaurus/src/webpack/react-dev-utils-webpack5/webpackHotDevClient.js

```diff
@@ -0,0 +1,285 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+/* eslint-disable */
+/*
+ * THIS FILE IS MODIFIED FOR DOCUSAURUS
+ * the above copyright header must be preserved for license compliance.
+ */
+
+// This alternative WebpackDevServer combines the functionality of:
+// https://github.com/webpack/webpack-dev-server/blob/webpack-1/client/index.js
+// https://github.com/webpack/webpack/blob/webpack-1/hot/dev-server.js
+
+// It only supports their simplest configuration (hot updates on same server).
+// It makes some opinionated choices on top, like adding a syntax error overlay
+// that looks similar to our console output. The error overlay is inspired by:
+// https://github.com/glenjamin/webpack-hot-middleware
+
```

Comment:
@RDIL for info I've reverted some of your changes here with the initial react-dev-utils code, as I think some changes were not necessary and we don't want to diverge too much from react-dev-utils code (and when we do, it should be documented)


Revert commit: https://github.com/facebook/docusaurus/pull/4089/commits/aa137b4218713480f219d9ce5319f317353db770

Let me know if you see anything important to keep here.

Your comment included: 

> Docusaurus's fork of this module adds webpack 5 support, makes it work with a typical D2 env, and improves perf.

I don't understand what you mean by "improves perf"